### PR TITLE
tests/nested/manual: extend preinstall check test to cover actions api

### PR DESCRIPTION
**Pull request by ernestl**
Add spread coverage for error actions

Error actions api was added [here](https://github.com/canonical/snapd/pull/15871)
Jira: https://warthogs.atlassian.net/browse/SNAPDENG-35035

Checking for the /run/mnt/.../preinstall file with the runchecks information for optimum PCR configuration will be added to a different general hybrid test running on 25.10 (in the works).

### CODE REVIEW

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml [hunk 1]
```diff
@@ -134,3 +136,49 @@ execute: |
...
+
+  # Check that encryption is again available for the targeted system
+  remote.exec "sudo snap debug api /v2/systems/classic 2>/dev/null" | gojq '.result["storage-encryption"].support' | MATCH "available"
+  remote.exec "sudo snap debug api /v2/systems/classic 2>/dev/null" | gojq '.result["storage-encryption"]."encryption-type"' | MATCH "cryptsetup"
+
+  # Restart snapd to clear preinstall context from cache
+  remote.exec "sudo systemctl restart snapd.service"
+
+  # Requesting a fix action when there is no preinstall context available should result in error
+  remote.exec "echo '{ \"action\": \"fix-encryption-support\", \"fix-action\": \"\"}' | sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems/classic 2>/dev/null" > result
```
Comment (bboozzoo) [inline tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: I think this will work too, but with less annoying quoting:
```suggestion
  echo '{"action": "fix-encryption-support", "fix-action": ""}' | \
      remote.exec "sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems/classic" > result
```

Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: yes that's much nicer, thanks

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml [hunk 2]
```diff
@@ -134,3 +136,56 @@ execute: |
...
+
+  # Requesting a fix action when there is no preinstall context available should result in error
+  echo '{"action": "fix-encryption-support", "fix-action": ""}' | remote.exec "sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems/classic 2>/dev/null" > result
+  cat <<EOF > result.reference
+  {
+    "result": {
+      "message": "cannot use check action without cached encryption information"
+    },
+    "status": "Internal Server Error",
+    "status-code": 500,
```
Comment (ZeyadYasser) [inline tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml:169]: I know this not related to the spread test, but shouldn't this have been a "Bad Request"?

Comment (ZeyadYasser) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml:169]: Since snapd seems to be doing the right thing

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml [hunk 3]
```diff
@@ -134,3 +136,56 @@ execute: |
...
+  EOF
+
+  diff result result.reference
+
+  # Run preinstall check to provide cached context
+  remote.exec "sudo snap debug api /v2/systems/classic 2>/dev/null" | gojq '.result["storage-encryption"].support' | MATCH "available"
+  remote.exec "sudo snap debug api /v2/systems/classic 2>/dev/null" | gojq '.result["storage-encryption"]."encryption-type"' | MATCH "cryptsetup"
+
+  # Requesting a fix action with preinstall context available should succeed
+  echo '{"action": "fix-encryption-support", "fix-action": ""}' |
```
Comment (ZeyadYasser) [inline tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: This is also not related to the spread test, but shouldn't the daemon fail for an empty `fix-action` or does an empty value have a special meaning?

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml [hunk 4]
```diff
@@ -134,3 +136,56 @@ execute: |
...
+  remote.exec "sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems/classic 2>/dev/null" |
```
Comment (ZeyadYasser) [inline tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: to avoid repetition and make it easier for reading the GET/POST calls could be wrapped in some function, like

https://github.com/canonical/snapd/blob/bab01ef5582040137e48fffd5caa36e64939de03/tests/nested/manual/passphrase-support-on-hybrid/task.yaml#L238-L243

**Review by ZeyadYasser — COMMENTED** [review]
Looks really good, only some tiny nitpick comments and questions

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 2 above)
Comment (bboozzoo) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml:169]: I think there was some major inconvenience with passing the enough context down and then back up to reply with 400

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml [hunk 5]
```diff
@@ -134,3 +136,56 @@ execute: |
...
+
+  # Check that encryption is again available for the targeted system
+  remote.exec "sudo snap debug api /v2/systems/classic 2>/dev/null" | gojq '.result["storage-encryption"].support' | MATCH "available"
+  remote.exec "sudo snap debug api /v2/systems/classic 2>/dev/null" | gojq '.result["storage-encryption"]."encryption-type"' | MATCH "cryptsetup"
+
+  # Restart snapd to clear preinstall context from cache
+  remote.exec "sudo systemctl restart snapd.service"
+
+  # Requesting a fix action when there is no preinstall context available should result in error
+  echo '{"action": "fix-encryption-support", "fix-action": ""}' | remote.exec "sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems/classic 2>/dev/null" > result
```
Comment (bboozzoo) [inline tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: I suppose you can skip the action completely instead of sending something empty?
```suggestion
  echo '{"action": "fix-encryption-support"}' | remote.exec "sudo snap debug api -X POST -H 'Content-Type: application/json' /v2/systems/classic 2>/dev/null" > result
```

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 2 above)
Comment (ZeyadYasser) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml:169]: aha, then it's no big deal, feel free to ignore

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 5 above)
Comment (ZeyadYasser) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: I had a similar concern https://github.com/canonical/snapd/pull/15996#discussion_r2354622849, it seems the fix-action is a `*string` so removing is not the same as an empty action, not sure if an empty action has a special meaning to it.

https://github.com/canonical/snapd/blob/bab01ef5582040137e48fffd5caa36e64939de03/daemon/api_systems.go#L746-L750

Comment (bboozzoo) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: maybe the check could be relaxed a bit? ActionNone sounds like something that would have no observable effects, so perhaps we could treat `null` and `""` the same

Comment (ZeyadYasser) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: if so, the string pointer can go away altogether, no?

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 3 above)
Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: well spotted and good question:

"" maps to secboot constant ActionNone with is always allowed, but inert.
Without other actions implemented atm, this is the only available action for happy scenario

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 5 above)
Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: its intentional, given its the action API it always requires an action otherwise it will complain. "" is a valid action from secboot perspective, and may even proof useful, so this maps 1-to-1 with secboot API.

Specifically changed to *string to differentiate no action from valid "" action...

secboot code:
image

Comment (bboozzoo) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: Looking at the code again. AFAIU  `action == ""` means the user chose no-action, which I understand the UI somehow needs to be made aware of and of the fact that this option is always available. Looking at https://github.com/canonical/secboot/blob/a9b8ee9246a941763217057c7f87360bab39d70b/efi/preinstall/checks_context.go#L197-L204 seems like it will be returned from snapd API? It's bit unfortunate that ActionNone is `""` which is a zero value of a string, which led to the problem at hand. I suppose there's no way around it now, unless we add an auxiliary value mapping ActionNone to something that's actually meaningful to the API consumer (e.g. "none"?). (to be clear I'm not suggesting we do it)

Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: I agree, "none" would be more convenient, @chrisccoulson what do you think can we still change this?

Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: for now i suppose we can keep it as is.

Comment (bboozzoo) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: the maybe add a comment in the test where `fix-action: ""` that `""` actually means no action

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 4 above)
Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: done

File: tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml (hunk 5 above)
Comment (ernestl) [reply tests/nested/manual/hybrid-tpm-fde-preinstall-check/task.yaml]: added comment

**Comment by ernestl** [comment]
Failures analysis:
- hybrid-tpm-fde-preinstall-check passed (runs in nested 24.04).
- failures do not relate to changes

**Review by bboozzoo — APPROVED** [review]
LGTM
